internal/v2/tls_config_store/tls_config_store.go

```diff
@@ -22,53 +24,188 @@ var (
+		log.Fatalf("Client: failed to send SessionReq: %v", err)
+	}
+	log.Printf("Client: sent SessionReq")
+
+	// Get the response containing config from S2Av2.
+	resp, err := cstream.Recv()
+	if err != nil {
+		log.Fatalf("Client: failed to recieve SessionResp: %v", err)
+	}
+	log.Printf("Client: recieved SessionResp")
+
+	// TODO(rmehta19): Handle resp.GetStatus().
+
+	// Extract TLS configiguration from SessionResp.
+	tlsConfig := resp.GetGetTlsConfigurationResp().GetClientTlsConfiguration()
+
+	// Fake S2Av2 only puts one cert in CertitificateChain, for loop iterates
+	// once.
+	var certList []tls.Certificate
+	for _, v := range tlsConfig.CertificateChain {
+		// TODO(rmehta19): Call remote signer library for private key.
+		cert, err := tls.X509KeyPair([]byte(v), clientKey)
+		if err != nil {
+			log.Fatalf("Failed to generate X509KeyPair: %v", err)
+		}
+		certList = append(certList, cert)
 	}
 
 	rootCertPool := x509.NewCertPool()
 	rootCertPool.AppendCertsFromPEM(serverCert)
 
-	// TODO(rmehta19): Call S2Av2 for config values.
+	// Map S2Av2 TLSVersion to consts defined in tls package.
+	var minVersion uint16
+	var maxVersion uint16
+	switch x := tlsConfig.MinTlsVersion; x {
+	case commonpb.TLSVersion_TLS_VERSION_1_0:
+		minVersion = tls.VersionTLS10
+	case commonpb.TLSVersion_TLS_VERSION_1_1:
+		minVersion = tls.VersionTLS11
+	case commonpb.TLSVersion_TLS_VERSION_1_2:
+		minVersion = tls.VersionTLS12
+	case commonpb.TLSVersion_TLS_VERSION_1_3:
+		minVersion = tls.VersionTLS13
+	default:
+		minVersion = tls.VersionTLS13
+	}
+
+	switch x := tlsConfig.MaxTlsVersion; x {
+	case commonpb.TLSVersion_TLS_VERSION_1_0:
+		maxVersion = tls.VersionTLS10
+	case commonpb.TLSVersion_TLS_VERSION_1_1:
+		maxVersion = tls.VersionTLS11
+	case commonpb.TLSVersion_TLS_VERSION_1_2:
+		maxVersion = tls.VersionTLS12
+	case commonpb.TLSVersion_TLS_VERSION_1_3:
+		maxVersion = tls.VersionTLS13
+	default:
+		maxVersion = tls.VersionTLS13
+	}
+
 	// Create mTLS credentials for client.
 	return &tls.Config {
-		Certificates: []tls.Certificate{cert},
+		// TODO(rmehta19): Make use of tlsConfig.HandshakeCiphersuites /
+		// RecordCiphersuites.
+		Certificates: certList,
 		VerifyPeerCertificate: certverifier.VerifyServerCertificateChain("s2a_test_cert", serverHostname, rootCertPool),
 		RootCAs: rootCertPool,
 		ServerName: serverHostname,
 		InsecureSkipVerify: true,
 		ClientSessionCache: nil,
-		MinVersion: uint16(tls.VersionTLS13),
-		MaxVersion: uint16(tls.VersionTLS13),
+		MinVersion: minVersion,
+		MaxVersion: maxVersion,
 	}
 }
 
 // GetTlsConfigurationForServer returns a tls.Config instance for use by a server application.
-func GetTlsConfigurationForServer() *tls.Config {
-	// TODO(rmehta19): Call S2Av2 for certificate.
-	// TODO(rmehta19): Call remote signer library for private key.
-	cert, err := tls.X509KeyPair(serverCert, serverKey)
+func GetTlsConfigurationForServer(cstream s2av2pb.S2AService_SetUpSessionClient) *tls.Config {
```

Comment:
When building the server-side tls.Config, we don't want to call S2A as soon as GetTlsConfigurationForServer is called. Instead, we want to create a helper function that will do so and use it to populate the tls.Config's GetConfigForClient field. We can make this change in a follow-up PR if you want, but please add a TODO as a reminder.

Reply:
Will do, I'll add a TODO and be sure to get to it in the next PR! This helper function will be something like 'sendServerTLSConfigRequest' and it will build the SessionReq, rather than doing that in the function itself? Do we want to do the same on the client side also then? Also, is the reasoning behind moving the cstream.Send(&SessionReq) to make the GetTlsConfiguration implementation simpler?

Reply:
Will do, I'll add a TODO and be sure to get to it in the next PR! This helper function will be something like 'sendServerTLSConfigRequest' and it will build the SessionReq, rather than doing that in the function itself? Do we want to do the same on the client side also then? Also, is the reasoning behind moving the cstream.Send(&SessionReq) to make the GetTlsConfiguration implementation simpler?